Assets/MixedRealityToolkit.SDK/Experimental/Features/Utilities/Migration/Tools/MigrationWindow.cs

```diff
@@ -0,0 +1,417 @@
+using UnityEngine;
+using Microsoft.MixedReality.Toolkit.Utilities.Editor;
+using UnityEngine.SceneManagement;
+using UnityEditor.SceneManagement;
+
+using Object = UnityEngine.Object;
+
+namespace Microsoft.MixedReality.Toolkit.Experimental.Utilities
+{
+    /// <summary>
+    /// This tool allows the migration of game object obsolete components into up-to-date versions.
+    /// Requires Implementation of IMigrationhandler.
+    /// </summary>
+    public class MigrationWindow : EditorWindow
+    {
+        private enum ToolbarOption 
+        { 
+            Objects = 0, 
+            Scenes = 1, 
+            Project = 2 
+        };
+
+        private static readonly string[] toolbarTitles =
+{
+            "Objects",
+            "Scenes",
+            "Full Project"
+        };
+
+        private ToolbarOption selectedToolbar = ToolbarOption.Objects;
+        private Vector2 scrollPosition = Vector2.zero;
+
+        private const string DependencyWindowURL = "https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Tools/MigrationWindow.html";
+        private const string WindowTitle = "Migration Window";
+        private const string WindowDescription = "This tool allows the migration of obsolete components into up-to-date versions. Component-specific implementation of the MigrationHandler Interface is required.";
+
+        private const string SceneExtension = ".unity";
+        private const string PrefabExtension = ".prefab";
+
+        private Dictionary<string, Type> migrationTypesMap = new Dictionary<string, Type>();
+        private String[] migrationTypeNames;
+        private int selectedMigrationHandlerIndex;
+        private List<Object> migrationObjects = new List<Object>();
+        private Object selection;
+
+        private dynamic migrationTypeInstance
+        {
+            get
+            {
+                var selectedMigrationTypeName = migrationTypeNames[selectedMigrationHandlerIndex];
+                var migrationHandlerType = migrationTypesMap[selectedMigrationTypeName];
+                return Activator.CreateInstance(migrationHandlerType) as IMigrationHandler;
+            }
+        }
+
+        [MenuItem("Mixed Reality Toolkit/Utilities/Migration Window", false, 4)]
+        private static void ShowWindow()
+        {
+            var window = GetWindow<MigrationWindow>(typeof(SceneView));
+            window.titleContent = new GUIContent(WindowTitle, EditorGUIUtility.IconContent("d_TimelineEditModeRippleOFF").image);
+            window.minSize = new Vector2(400.0f, 300.0f);
+            window.Show();
+        }
+
+        private void OnGUI()
+        {
+            if (EditorApplication.isPlaying || EditorApplication.isPaused)
+            {
+                GUI.enabled = false;
+            }
+            DrawHeader();
+            DrawMigrationTypeSelector();
+            DrawMigrationToolbars();
+        }
+
+        private void OnFocus()
+        {
+            RefreshAvailableTypes();
```

Comment:
If it isn't too hard to do, it might also be good to gate this on recompiles as well (or asset changes) - i.e. arbitrary focus changes maybe this isn't super interesting.

HOWEVER if you're using this and this is fairly fast, then you don't need to do this - it's just if this adds micro hangs that we'd want to do this. I don't want to ask for premature optimization.

Reply:
was my initial thought... looking into it